crates/service-client/src/http.rs

```diff
@@ -225,16 +228,50 @@ impl HttpError {
         // HTTP1.1 method, so typically 1.1 servers respond with a 40x, and the h2 client sees
         // this as an invalid frame.
         err.source()
-            .and_then(|err| err.downcast_ref::<h2::Error>())
+            // Cause can either be h2 directly, or hyper::Error and then h2.
+            .and_then(|err| {
+                err.downcast_ref::<h2::Error>().or_else(|| {
+                    err.downcast_ref::<hyper::Error>()
+                        .and_then(|e| e.source())
+                        .and_then(|err| err.downcast_ref::<h2::Error>())
+                })
+            })
             .and_then(|err| err.reason())
             == Some(h2::Reason::FRAME_SIZE_ERROR)
     }
+
+    fn is_possible_h2_only_error(err: &hyper_util::client::legacy::Error) -> bool {
+        // This is a reasonably fuzzy check to figure out if the user passed --http1.1-only when doing discovery,
+        // but the service supports only HTTP/2
+        err.source()
+            .and_then(|err| err.downcast_ref::<hyper::Error>())
+            .map(|err| {
+                use std::fmt::Write;
+
+                struct Matcher(bool);
+                impl Write for Matcher {
+                    fn write_str(&mut self, s: &str) -> fmt::Result {
+                        if s == "invalid HTTP version parsed" {
+                            self.0 = true;
+                        }
+                        Ok(())
+                    }
+                }
+
+                let mut matcher = Matcher(false);
+                let _ = write!(&mut matcher, "{err}");
```

Comment:
Are we going through a `Write` implementation to avoid the String allocation of `err.to_string()` and then doing a `.contains()`? Wondering whether this kind of cleverness is really needed (at the cost of less obvious code). It could deserve a comment. Alternatively, one could just iterate through the sources if things should be explicited.